src/cls/queue/cls_queue_src.cc

```diff
@@ -0,0 +1,478 @@
+
+  bool offset_populated = false, entry_start_processed = false;
+  uint64_t data_size = 0, num_ops = 0;
+  uint16_t entry_start = 0;
+  bufferlist bl;
+  do
+  {
+    CLS_LOG(10, "INFO: queue_list_entries(): start_offset is %lu\n", start_offset);
+  
+    bufferlist bl_chunk;
+    //Read chunk size at a time, if it is less than contiguous data size, else read contiguous data size
+    if (contiguous_data_size > chunk_size) {
+      size_to_read = chunk_size;
+    } else {
+      size_to_read = contiguous_data_size;
+    }
+    CLS_LOG(10, "INFO: queue_list_entries(): size_to_read is %lu\n", size_to_read);
+    if (size_to_read == 0) {
+      next_marker = head.tail;
+      op_ret.is_truncated = false;
+      CLS_LOG(20, "INFO: queue_list_entries(): size_to_read is 0, hence breaking out!\n");
+      break;
+    }
+
+    auto ret = cls_cxx_read(hctx, start_offset, size_to_read, &bl_chunk);
+    if (ret < 0) {
+      return ret;
+    }
+
+    //If there is leftover data from previous iteration, append new data to leftover data
+    bl.claim_append(bl_chunk);
+    bl_chunk = bl;
+    bl.clear();
+
+    CLS_LOG(20, "INFO: queue_list_entries(): size of chunk %u\n", bl_chunk.length());
+
+    //Process the chunk of data read
+    unsigned index = 0;
+    auto it = bl_chunk.cbegin();
+    uint64_t size_to_process = bl_chunk.length();
+    do {
+      CLS_LOG(10, "INFO: queue_list_entries(): index: %u, size_to_process: %lu\n", index, size_to_process);
+      cls_queue_entry entry;
+      it.seek(index);
+      //Populate offset if not done in previous iteration
+      if (! offset_populated) {
+        cls_queue_marker marker = {start_offset + index, gen};
+        CLS_LOG(5, "INFO: queue_list_entries(): offset: %s\n", marker.to_str().c_str());
+        entry.marker = marker.to_str();
+      }
+      // Magic number + Data size - process if not done in previous iteration
+      if (! entry_start_processed ) {
+        if (size_to_process >= (sizeof(uint16_t) + sizeof(uint64_t))) {
+          // Decode magic number at start
+          try {
+            decode(entry_start, it);
+          } catch (buffer::error& err) {
+            CLS_LOG(10, "ERROR: queue_list_entries: failed to decode entry start \n");
+            return -EINVAL;
+          }
+          if (entry_start != QUEUE_ENTRY_START) {
+            CLS_LOG(5, "ERROR: queue_list_entries: invalid entry start %u\n", entry_start);
+            return -EINVAL;
+          }
+          index += sizeof(uint16_t);
+          size_to_process -= sizeof(uint16_t);
+          // Decode data size
+          try {
+            decode(data_size, it);
+          } catch (buffer::error& err) {
+            CLS_LOG(10, "ERROR: queue_list_entries: failed to decode data size \n");
+            return -EINVAL;
+          }
+        } else {
+          // Copy unprocessed data to bl
+          bl_chunk.copy(index, size_to_process, bl);
```

Comment:
@cbodley : I checked the method declaration of copy and can see that it is 'horribly inefficient' for large bufferlists', the chunk size is 1024 bytes so can we use the copy method here or should I look at other bufferlist methods?

Reply:
its only 'horribly inefficient' when the bufferlist has a lot of segments, because it has linear complexity for the seek to index, and for the copying

but you don't really need to copy the data here right? you just want to transfer the leftover bytes back to bl. you can use bufferlist::splice() with the same arguments to do that

Reply:
Yeah ok